content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+
+## Extension Design
+
+The Type enum in the [bril_rs] library is modified as follows:
+
+```rust
+// I've removed the parts of this enum used in serializing/deserializing to json.
+pub enum Type {
+    Int,
+    Bool,
+    Float,
+    Pointer(Box<Type>),
+
+    /// New addition
+    PointerRegions {
+        pointer_type : Box<Type>,
+        ownership : Option<Ownership>,
+        region: String,
+    },
+    ///
+}
+```
+
+The new addition is a pointer type which is annotated with the region it exists in and a concept of whether a pointer owns the memory it is pointing to. In Bril, every ```id```, ```load```, ```alloc```, and ```ptr_add``` can create a new pointer into memory. However, not all pointers are created equal. Pointer's
+created by ```alloc``` are the original pointers into the zeroth index of memory and are treated like the "Owners" of that memory. It is these pointers that need to be freed when the scope of their region ends. Pointers created by ```id```, ```load```, and ```ptr_add``` are merely memory borrowers with views into heap memory. They still need to be statically checked that they aren't being used after the region scope ends but do not need a corresponding ```free``` instruction like memory owners do. What is important about this typing is that both pointer types can exist within a program depending on whether you want the region to be inferred or explicitly stated. Whether a pointer owns its memory is trivial to decide and it is not expected that programmers included this annotation. As will be touched on in the evaluation section, this opt-in use of explicit regions makes converting valid programs to this extension simple.
+
+## Current Limitations on Programs
+
+Outside of the type system additions, there are a few restriction put onto Bril Programs for them to be valid for this extension:
+
+- Any variable that owns it's memory(returned from ```alloc```) can not be overwritten during the course of the program. This is to ensure that a program doesn't make multiple allocations using the same variable name but whereas the static analysis can only free the variables once.
+- Currently, programmers cannot declare smaller regions within the function scope. Currently, each function has a region and each pointer argument has it's own region. The next step would be to allow the programmer to create their own scoped regions. This was set asider during the project proposal process but would be interesting further work.
```

Comment:
"its" (I'm gong to stop pointing out it's/its errors now… please audit them throughout.)